### PR TITLE
Add radius arc

### PR DESCRIPTION
**Pull request by gntech**
Add an function that draws an arc that is determined by startPoint, endPoint and radius.

### CODE REVIEW

**Comment by dcowden** [comment]
@gntech thanks for another contribution! this looks great!
Is it ready to merge in from your view?

**Comment by gntech** [comment]
Actually, I was thinking about some things.

1. For example: If the linear distance between startPoint and endPoint is more than 2 * radius the arc will fail since it it geometrically impossible. If this is the case I would like to raise some kind of Error and 
present it to the user in a logical way. Do you already do this kind of sanity checking of the parameters for other functions? Can you point me to some example?

```python
if endPoint.sub(startPoint).Length > 2 * radius:
    raise ValueError
```

2. The tests I added for the radiusArc and sagittaArc functions are basically just a test that the arc functions dont throw any error. They don't actually test if the result is correct. Do you have any suggestion how I can improve the test cases for the arc functions?

**Comment by jmwright** [comment]
> Do you already do this kind of sanity checking of the parameters for other functions? Can you point me to some example?

I think that rasing a ValueError is how we do it elsewhere. See [here](https://github.com/dcowden/cadquery/blob/8e20d7d8f17a5c0db94e7dbc06bff4eefc924cd2/cadquery/freecad_impl/geom.py#L345) for a case that's inside the CQ core code, and [here](https://github.com/dcowden/cadquery/blob/22fabda8d8de63ae19d82e34934c29b5f29940c1/examples/FreeCAD/Ex029_Braille.py#L182) for a case that's in an example.

> Do you have any suggestion how I can improve the test cases for the arc functions?

Not sure if this will do exactly what you're looking for, but you can check the length of edges. The last line in this sample code is the important one.

```python
import cadquery as cq

(L, H, W, t) = (100.0, 20.0, 20.0, 1.0)

# Define the points that the polyline will be drawn to/thru
pts = [
    (W/2.0, H/2.0),
    (W/2.0, (H/2.0 - t)),
    (t/2.0, (H/2.0-t)),
    (t/2.0, (t - H/2.0)),
    (W/2.0, (t - H/2.0)),
    (W/2.0, H/-2.0),
    (0, H/-2.0)
]

result = cq.Workplane("front").moveTo(0, H/2.0) \
                              .polyline(pts)

# Do check on edge lengths here. May have to get the radius indirectly
print(result.edges('>Y').first().val().Length())
```
Notice the `.val()` call to get the underlying shape object. The higher level Workplane object doesn't have functions like `Length()` attached to it.

**Comment by jmwright** [comment]
@gntech Also, after looking at the website you have in your profile link, you might want to check out issue #173 @RustyVermeer has been working on DXF import, which may help you a lot, especially if the 2D profiles in your designs get complex.

**Comment by fragmuffin** [comment]
I would create a series of arcs and assert:

* start, end, and midpoint locations
* wire type (`edge.geomType() == 'CIRCLE'`)

This [apply_spline() function](https://github.com/fragmuffin/cqparts/blob/master/src/cqparts_fasteners/solidtypes/threads/base.py#L151-L165) iterates along an arc or a spline getting vertices at different lengths along the wire.
It shows how to get the length of an arc.

File: cadquery/cq.py [hunk 1]
```diff
@@ -1366,6 +1366,34 @@ def sagittaArc(self, endPoint, sag, forConstruction=False):
...
+        A positive radius means convex arc and negative radius means concave arc.
+        """
+
+        startPoint = self._findFromPoint(False)
+        endPoint = self.plane.toWorldCoords(endPoint)
+        midPoint = endPoint.add(startPoint).multiply(0.5)
+
+        # Calculate the sagitta from the radius
+        length = endPoint.sub(startPoint).Length / 2.0
+        sag = abs(radius) - math.sqrt(radius**2 - length**2)
```
Comment (fragmuffin) [inline cadquery/cq.py]: As you mentioned, if the `radius < length` then the arc is not possible.
I think the best way to capture this is with a `try` block over the `sag` calculation

```python
try:
    sag = abs(radius) - math.sqrt(radius**2 - length**2)
except ValueError:
    raise ValueError("Arc radius is not large enough to reach the given end-point")
```

Comment (gntech) [reply cadquery/cq.py]: @jmwright and @fragmuffin: Thanks! Good suggestions, I will try to get this implemented later this weekend.

**Comment by gntech** [comment]
See my latest commit, I discovered something strange with the toWorldCoords function. When runned in sequence it changes the coordinates. I dont think that this is the intent?

**Comment by gntech** [comment]
I realize that I made some wrong assumptions about what toWorldCoords do.

But this is my intention: sagittaArc, radiusArc and threePointArc all uses startPoint and endPoint but determines the arc between them in different ways: sag, radius or midPoint.

My idea is that sagittaArc and radiusArc should just calculate the corresponding midPoint and then hand that midpoint over to threePointArc for the actual arc creation.

Does this sound reasonable or is it prefferred to have separate functions that create the arcs by themself?

I will try to write some code later tonight to better explain.

**Comment by jmwright** [comment]
> Does this sound reasonable or is it prefferred to have separate functions that create the arcs by themself?

I don't know what @dcowden thinks, but I think your approach is fine. I try to reuse existing building blocks whenever possible, which it sounds like you're doing.

**Comment by dcowden** [comment]
@gntech i like your approach.

**Comment by dcowden** [comment]
@gntech this looks good, +1 from me. Are you ready to merge?

@jmwright does this get +1 from you?

**Comment by gntech** [comment]
Yes, I consider this ready for merge

**Comment by jmwright** [comment]
+1 from me

**Comment by dcowden** [comment]
@gntech thanks for this contribution! You rock!

**Comment by gntech** [comment]
Thank you! And thanks for your work on cadquery which is awesome!

**Comment by dcowden** [comment]
@gntech absolutely! thanks for the kind words.

I think we can do really amazing things with cq once we get an OCC layer implemented, and we get proper feature trees and such.

**Pull request merged**